File: clients/typescript/client/src/execution.ts
import { FyndClientError } from "./error.js";

export interface SettledOrder {
  txHash: string;
  /** Amount of output token actually received, from on-chain Transfer logs. */
  amountReceived: bigint;
  blockNumber: number;
}

/** ERC-20 Transfer event topic: keccak256("Transfer(address,address,uint256)") */
const TRANSFER_TOPIC =
  "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef";

export interface EthLog {
  address: string;
  topics: string[];
  data: string;
  blockNumber: string | number;
}

export interface TxReceipt {
  status: string; // "0x1" = success
  blockNumber: string | number;
  logs: EthLog[];
}

/**
 * Derives the amount received from on-chain Transfer event logs.
 *
 * Looks for ERC-20 Transfer events from `tokenOut` to `receiver`
 * and sums their values from log data.
 */
export function extractAmountReceived(
  receipt: TxReceipt,
  tokenOut: string,
  receiver: string,
): bigint {
  const tokenOutLower = tokenOut.toLowerCase();
  const receiverLower = receiver.toLowerCase().replace(/^0x/, "").padStart(40, "0");

  return receipt.logs
    .filter((log) => {
      if (log.address.toLowerCase() !== tokenOutLower) return false;
      if (log.topics[0]?.toLowerCase() !== TRANSFER_TOPIC) return false;
      // topics[2] is the indexed `to` address (padded to 32 bytes)
      const toTopic = log.topics[2]?.toLowerCase().slice(-40);
      return toTopic === receiverLower;
    })
    .reduce((acc, log) => {
      // Transfer value is the non-indexed data field (first 32 bytes = 256 bits)
      const value = BigInt("0x" + log.data.replace(/^0x/, "").slice(0, 64));
      return acc + value;
    }, 0n);
}

/**
 * A handle to a submitted transaction.
 * Call `settle(rpcUrl)` to wait for confirmation and parse logs.
 */
export class TransactionHandle {
  constructor(
    public readonly txHash: string,
    private readonly tokenOut: string,
    private readonly receiver: string,
  ) {}

  /**
   * Polls for the transaction receipt and returns a settled order.
   *
   * Uses eth_getTransactionReceipt via the provided RPC URL.
   * Derives amountReceived from ERC-20 Transfer logs — no debug RPC required.
   */
  async settle(rpcUrl: string, timeoutMs = 120_000): Promise<SettledOrder> {
    const deadline = Date.now() + timeoutMs;

    while (Date.now() < deadline) {
      const receipt = await this.getReceipt(rpcUrl);

      if (receipt !== null) {
        if (receipt.status !== "0x1") {
          throw new FyndClientError("transaction reverted", false);
        }

        const blockNumber =
          typeof receipt.blockNumber === "string"
            ? parseInt(receipt.blockNumber, 16)
            : receipt.blockNumber;

        const amountReceived = extractAmountReceived(
          receipt,
          this.tokenOut,
          this.receiver,
        );

        return { txHash: this.txHash, amountReceived, blockNumber };
      }

      await new Promise((r) => setTimeout(r, 2000));
    }

    throw new FyndClientError(
      `transaction ${this.txHash} not confirmed within ${timeoutMs}ms`,
      true,
    );
  }

  private async getReceipt(rpcUrl: string): Promise<TxReceipt | null> {
    const response = await fetch(rpcUrl, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        jsonrpc: "2.0",
        id: 1,
        method: "eth_getTransactionReceipt",
        params: [this.txHash],
      }),
    });

    if (!response.ok) {
      throw new FyndClientError(
        `RPC request failed: ${response.status}`,
        response.status >= 500,
      );
    }

    const json = (await response.json()) as {
      result: TxReceipt | null;
      error?: { message: string };
    };
    if (json.error) {
      throw new FyndClientError(`RPC error: ${json.error.message}`, false);
    }
    return json.result;
  }
}

export type ExecutionReceipt =
  | { type: "transaction"; handle: TransactionHandle }
  | { type: "intent" }; // Turbine placeholder

File: clients/rust/src/execution.rs
use std::sync::Arc;

use alloy::network::Ethereum;
use alloy::primitives::{Address, TxHash, U256};
use alloy::providers::Provider;

use crate::error::FyndClientError;

/// A handle to a submitted transaction, resolves to a `SettledOrder`.
#[derive(Clone)]
pub struct TransactionHandle {
    pub(crate) tx_hash: TxHash,
    pub(crate) provider: Arc<dyn Provider<Ethereum>>,
    pub(crate) token_out: Address,
    pub(crate) receiver: Address,
}

/// An order that has been confirmed on-chain.
#[derive(Debug, Clone)]
pub struct SettledOrder {
    pub tx_hash: TxHash,
    /// Amount of output token actually received, derived from on-chain Transfer logs.
    pub amount_received: U256,
    pub block_number: u64,
}

/// Receipt returned after broadcasting a signed order.
pub enum ExecutionReceipt {
    /// Transaction was submitted; poll `settle()` to wait for confirmation.
    Transaction(TransactionHandle),
    /// Turbine intent — not yet implemented.
    Intent,
}

// keccak256("Transfer(address,address,uint256)")
const TRANSFER_TOPIC: alloy::primitives::B256 =
    alloy::primitives::b256!("ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef");

impl TransactionHandle {
    /// Waits for the transaction to be confirmed and returns the settled order.
    ///
    /// Derives `amount_received` from on-chain ERC-20 Transfer event logs —
    /// no debug RPC required.
    pub async fn settle(self) -> Result<SettledOrder, FyndClientError> {
        let receipt = self
            .provider
            .get_transaction_receipt(self.tx_hash)
            .await
            .map_err(|e| FyndClientError::Rpc(e.to_string()))?
            .ok_or_else(|| FyndClientError::Rpc("transaction not found".to_string()))?;

        let block_number = receipt
            .block_number
            .ok_or_else(|| FyndClientError::Rpc("receipt missing block number".to_string()))?;

        let amount_received = receipt
            .inner
            .logs()
            .iter()
            .filter(|log| {
                log.address() == self.token_out
                    && log.topics().first() == Some(&TRANSFER_TOPIC)
                    && log
                        .topics()
                        .get(2)
                        .map(|t| {
                            // indexed `to` address is padded to 32 bytes
                            let addr_bytes = &t.0[12..];
                            addr_bytes == self.receiver.as_slice()
                        })
                        .unwrap_or(false)
            })
            .filter_map(|log| {
                // Transfer value is in data (non-indexed)
                let data = log.data().data.as_ref();
                if data.len() >= 32 {
                    Some(U256::from_be_slice(&data[..32]))
                } else {
                    None
                }
            })
            .fold(U256::ZERO, |acc, v| acc + v);

        Ok(SettledOrder { tx_hash: self.tx_hash, amount_received, block_number })
    }
}
